### PR TITLE
Openstack support

### PR DESCRIPTION
**Pull request by cbaenziger**
This seems to be the necessary bits for me to stand up an OpenStack ([Chef-BCPC](https://github.com/bloomberg/chef-bcpc)) based Chef-BACH cluster.

One uses:
- The OpenStack Vagrant Provider
  - For Neutron based OpenStack clouds: https://github.com/cloudbau/vagrant-openstack-plugin
  - For Nova-Networking based OpenStack clouds: https://github.com/cbaenziger/vagrant-openstack-plugin/tree/0.8.0-volume-support
- CACert Plugin: https://github.com/williambailey/vagrant-ca-certificates.git

One provides in the Vagrant file:
- Floating IPs
- "Cluster Name" (string to prepend on VM names, if creating multiple clusters in the same OpenStack tenant)
- Flavor
- Image
- Keypair Name
  - The ssh key should be provided in `vagrant_ssh/id_rsa`
- SSL Certificates to Load 

Ensuring the following environment variables are set:
- `OS_USERNAME`
- `OS_PASSWORD`
- `OS_TENANT_NAME`
- `OS_AUTH_URL`
- Creates cluster VM's: `vagrant up`
- Create bootstrap server: `vagrant up <cluster name>bcpcchef`

Remaining to be resolved to not break hardware and VBox clusters:
- environment changes (impetus for issue #61)
- `kepalived` should still be in the head-node role
- Make a bit better sense of the WIP tree of over 25 commits

### CODE REVIEW

File: Vagrantfile.openstack [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    bcpcchef.vm.hostname = fqdn
+    bcpcchef.vm.provider :openstack do |os|
+     vm_settings.call(os)
+     os.server_name = hostname
+     os.floating_ip = FLOATING_IPS[0]
+     os.flavor = /m1.medium/
+    end
+
+    # spoof having a vagrant homedir
+    bcpcchef.vm.provision :shell, :inline => "ln -s /home/ubuntu /home/vagrant ; ln -s /vagrant /home/ubuntu/chef-bcpc ; chown -R ubuntu /home/ubuntu/"
```
Comment (kiiranh) [inline Vagrantfile.openstack]: This and following vm.provision statements would read better in a ruby-block-style provision block.

File: environments/Test-Laptop.json [hunk 2]
```diff
@@ -63,7 +41,6 @@
        "servers" : [ "0.pool.ntp.org", "1.pool.ntp.org", "2.pool.ntp.org", "3.pool.ntp.org" ]
     },
     "chef_client": {
-      "server_url": "http://10.0.100.3:4000",
```
Comment (kiiranh) [inline environments/Test-Laptop.json]: I assume the OpenStack code will reside in a separate branch in chef-bach? If not how does one handle merging the environment file with the master version?

File: cookbooks/bcpc-hadoop/recipes/hadoop_config.rb [hunk 3]
```diff
@@ -63,3 +63,8 @@
 package "openjdk-7-jdk" do
     action :upgrade
 end
+
+dpkg_autostart "hadoop-mapreduce" do
+  allow false
+end
+package "hadoop-mapreduce"
```
Comment (kiiranh) [inline cookbooks/bcpc-hadoop/recipes/hadoop_config.rb]: Isn't `action :upgrade` a better option for the package?

File: Vagrantfile.openstack [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+if json_file.empty?
+  puts "No environment file found to parse. Please make sure at least one environment file exists."
+  exit
+end
+if json_file.length > 1
+  puts "More than one environment file found."
+  exit
+end
+
+file_name=File.basename(json_file.join(","))
```
Comment (kiiranh) [inline Vagrantfile.openstack]: `file_name` is never used.

File: Vagrantfile.openstack [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    # Set a bogus address incase we do not have a bcpc-1 VM at this time
+    vm1_fixed_ip = "## DID NOT FIND #{CLUSTER_NAME}bcpc-1 VM ##"
+    if not id
+      nil
+      puts "Unable to find host #{CLUSTER_NAME}bcpc-1 found #{machines.length()} matches"
+    else
+      ips = openstack_compute.servers.get(id).private_ip_addresses
+      if not ips.length() == 1
+        nil
+        puts "Unable to find IP for #{CLUSTER_NAME}bcpc-#{idx} found #{ips.length()} IPs"
```
Comment (kiiranh) [inline Vagrantfile.openstack]: Typo here... `#{idx}` should be `#{id}`

File: cookbooks/bcpc/recipes/networking.rb [hunk 6]
```diff
@@ -118,110 +118,93 @@
 bash "setup-interfaces-source" do
   user "root"
   code <<-EOH
-    echo "source /etc/network/interfaces.d/iface-*" >> /etc/network/interfaces
+    echo "source /etc/network/interfaces.d/*.cfg" >> /etc/network/interfaces
   EOH
-  not_if "grep '^source /etc/network/interfaces.d/' /etc/network/interfaces"
+  not_if "grep '^source /etc/network/interfaces.d/\*.cfg' /etc/network/interfaces"
```
Comment (kiiranh) [inline cookbooks/bcpc/recipes/networking.rb]: In my OpenStack cluster, I ran cluster-assign-roles 3 times and I see 3 entries for `source /etc/network/interfaces.d/\*.cfg` in `/etc/network/interfaces`. However, when I execute the not_if code on the shell, the grep does work. Not sure whats wrong.

Comment (cbaenziger) [inline cookbooks/bcpc/recipes/networking.rb]: This has been fixed upstream at this point.

**Comment by vt0r** [comment]
Closing for now. Feel free to reopen if this work gets picked up later on.

**Pull request closed without merging**